### DISCUSSION TITLE
Installing VSCode Extension

### DISCUSSION THREAD

**Discussion by Chagrilled** [Homebrew/discussions#294] [Everyday usage]
Hey,

Trying to write a formula to install a company-private VSCode extension. I've gotten to the point where I can get VSCode to "do something" in that it pops up in the corner of my Mac, then silently disappears again, with no output anywhere, alas no extension is installed. 

Writing Brew formulae (and ruby) is pretty new to me, but I expect there to be some sort of permission error possibly? Hard to tell. If I run the command in my own shell, it's fine.
 
Here's my formula install step. Previously I used `code --install-extension` but I presume `code` is not on the $path of whoever brew runs as. Using backticks, I got some output from `--install-extension` in which it expectedly moaned about me using Node `Buffer` classes, which means it actually has read the contents of the extension bundle, but it's still not installed.

```ruby
class VscodeExt < Formula
  desc ""
  homepage ""
  url "https://redacted/vscode-ext-0.0.5.vsix"
  sha256 "..."
  license ""

def install
    system "/usr/local/bin/code --install-extension $(brew --cache vscode-ext)" # or below
    `/usr/local/bin/code --install-extension $(/usr/local/bin/brew --cache vscode-ext)`
    File.write("#{buildpath}/vscode-ext.txt", "this is a test file")
    bin.install "#{buildpath}/vscode-ext.txt" #filler file to avoid brew empty install errors
  end

...
```

I'm not sure if this is the right place for this, it might be more of a VSCode thing. Any ideas of what's going wrong here?

**Comment by carlocab** [comment]
What is `vscode-ext`? Is that a separate formula you have in a tap? (It's not the formula that contains this `install` block, is it?)

Hard to really diagnose what the problem is without knowing what the rest (or at least, the earlier parts) of the formula looks like, as they determine the behaviour of the `install` block.

**Reply by Chagrilled** [reply]
Should have specified - it's the formula from which this install block is in. My URL points to a `.vsix` which VSCode ingests to install the extension. I ended up doing `brew --cache` to get the path of that file to point VSCode at to install. 

I'm actually now theorising that extension installations are per-user, and not system-wide.
I've edited the OP to include the rest of the formula, but it really is just downloading an artefact then installing it. I don't have VSCode installed via Brew on my own machine, so it's being "depended on", as I think one can.

**Reply by carlocab** [reply]
That's a strange way of finding the `.vsix` file, as whatever the URL points to will be unpacked and placed in `buildpath`, which is also the working directory of the `install` block. Did something like `code --install-extension vscode-ext-0.0.5.vsix` not work?

It may well be a permission error. I _think_ `brew install` runs things in a sandbox, which limits your ability to write anywhere outside of `brew --cellar formula` (along with `buildpath`, and possibly its parent), where `formula` is the formula being installed. Not 100% sure though. Maybe it'll work in a `post-install` block?

**Reply by Chagrilled** [reply]
`#{buildpath}` does 'work' now, I guess I was moving multiple things around at once to figure something out. I think the writing is ok, since it's calling `code` which then does the writing on its behalf with the right perms. I'm not sure what user brew runs as, but I'm suspecting it's installing correctly as that user, since extensions and their storage are actually stored under the home directory. 

I suppose my only solution in this avenue is to try have brew execute the command as our actual user.

**Reply by carlocab** [reply]
Sounds tricky, but I don't have an alternative to offer you. However, here's a tip so that your formula also works for non-standard Homebrew installations that don't use `/usr/local` as its prefix (e.g. `brew` on Apple Silicon):
```
system "#{HOMEBREW_PREFIX}/bin/code", "--install-extension", "$(#{HOMEBREW_PREFIX}/bin/brew --cache vscode-ext)"
```
I'm starting to think though that `brew` might not be the right tool to use for your use case. I think it'd be simpler to just write a shell script.

**Reply by Chagrilled** [reply]
Cheers for the help, I'm thinking I might head over to the VSCode camp and ask there.

I currently have an install shell script on the repo, but was hoping to have a magical one-click `brew install extension-name` to avoid people needing to clone it.

**Reply by carlocab** [reply]
There's always
```
curl URL/of/raw/script | sh
```
but I don't know if you want to encourage the habit of piping `curl` into `sh`.
